compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala

```diff
@@ -0,0 +1,281 @@
+import coverage.*
+
+/** Implements code coverage by inserting calls to scala.runtime.Invoker
+  * ("instruments" the source code).
+  * The result can then be consumed by the Scoverage tool.
+  */
+class InstrumentCoverage extends MacroTransform with IdentityDenotTransformer:
+  import ast.tpd._
+
+  override def phaseName = InstrumentCoverage.name
+
+  override def description = InstrumentCoverage.description
+
+  // Enabled by argument "-coverage OUTPUT_DIR"
+  override def isEnabled(using ctx: Context) =
+    ctx.settings.coverageOutputDir.value.nonEmpty
+
+  // Atomic counter used for assignation of IDs to difference statements
+  private val statementId = AtomicInteger(0)
+
+  private var outputPath = ""
+
+  // Main class used to store all instrumented statements
+  private val coverage = Coverage()
+
+  override def run(using ctx: Context): Unit =
+    outputPath = ctx.settings.coverageOutputDir.value
+
+    // Ensure the dir exists
+    val dataDir = new File(outputPath)
+    val newlyCreated = dataDir.mkdirs()
+
+    if !newlyCreated then
+      // If the directory existed before, let's clean it up.
+      dataDir.listFiles.nn
+        .filter(_.nn.getName.nn.startsWith("scoverage"))
+        .foreach(_.nn.delete())
+    end if
+    super.run
+
+    Serializer.serialize(coverage, outputPath, ctx.settings.coverageSourceroot.value)
+
+  override protected def newTransformer(using Context) = CoverageTransormer()
+
+  /** Transforms trees to insert calls to Invoker.invoked to compute the coverage when the code is called */
+  private class CoverageTransormer extends Transformer:
+    private val IgnoreLiterals = new Property.Key[Boolean]
+
+    private def ignoreLiteralsContext(using ctx: Context): Context =
+      ctx.fresh.setProperty(IgnoreLiterals, true)
+
+    override def transform(tree: Tree)(using ctx: Context): Tree =
+      tree match
+        // simple cases
+        case tree: (Import | Export | This | Super | New) => tree
+        case tree if (tree.isEmpty || tree.isType) => tree // empty Thicket, Ident, TypTree, ...
+
+        // Literals must be instrumented (at least) when returned by a def,
+        // otherwise `def d = "literal"` is not covered when called from a test.
+        // They can be left untouched when passed in a parameter of an Apply.
+        case tree: Literal =>
+          if ctx.property(IgnoreLiterals).contains(true) then
+            tree
+          else
+            instrument(tree)
+
+        // branches
+        case tree: If =>
+          cpy.If(tree)(
+            cond = transform(tree.cond),
+            thenp = instrument(transform(tree.thenp), branch = true),
+            elsep = instrument(transform(tree.elsep), branch = true)
+          )
+        case tree: Try =>
+          cpy.Try(tree)(
+            expr = instrument(transform(tree.expr), branch = true),
+            cases = instrumentCases(tree.cases),
+            finalizer = instrument(transform(tree.finalizer), true)
```

Comment:
I don't know what the `branch` argument does, but does the finalizer count as a branch since it will always be run? I see the Scala 2 plugin does the same so maybe there's a good reason for it: https://github.com/scoverage/scalac-scoverage-plugin/blob/main/scalac-scoverage-plugin/src/main/scala/scoverage/plugin.scala#L836

Reply:
`branch = true` marks the statement as a branch in the scoverage report. This is used by some tools to show partially covered branches in a different color. I think it can make sense for try/finally: if the program crashes you'd want to show that the whole try hasn't been entirely covered.